### DISCUSSION TITLE
Deep linking with Expo and Supabase

### DISCUSSION THREAD

**Discussion by javierguzman** [supabase/supabase#10754] [Questions] [answered]
Hello all!

I am making a React Native application with Expo and I would like to redirect users after sign up by email. However, I have not been able to achieve so and on Supabase Discord I have been told it is best to ask around here as they could not see what it is wrong.

The specific problem is that if I open expo on my mobile phone and I sign up. When I open the Supabase sign up email and click on the link, the app is open back but it is not redirected to the correct screen. It stays in the same screen as it was previously (the one saying about checking your email)

The code for sign up is as follows:
```
const redirectURL = Linking.createURL('/AccountVerified');
    const { error } = await supabase.auth.signUp({whatever, options: {
        emailRedirectTo: redirectURL
}}
```

Linking configuration is:
```
const prefix = Linking.createURL('/'); 
const linking: LinkingOptions<RootStackParamList> = {
  prefixes: [prefix],
  config: {
    screens: {
      Landing: 'Landing',
      AccountVerified: 'AccountVerified'
    }
  }
};
```
App with navigator is:
```
const prefix = Linking.createURL('/'); 
const linking: LinkingOptions<RootStackParamList> = {
  prefixes: [prefix],
  config: {
    screens: {
      Landing: 'Landing',
      AccountVerified: 'AccountVerified'
    }
  }
};

export default function App() {
  return (
    <LinearGradient colors={['#dc3761', '#fd774e']} style={styles.container}>
      <NavigationContainer linking={linking} fallback={<Text>Loading...</Text>}>
        <Stack.Navigator
          initialRouteName="Landing"
          screenOptions={{ headerTransparent: true }}
        >
          <Stack.Screen
            name="Landing"
            component={LandingScreen}
            options={{ headerShown: false }}
          />
          <Stack.Screen
            name="Login"
            component={LoginScreen}
            options={{ headerShown: false }}
          />
          <Stack.Screen
            name="Signup"
            component={SignupScreen}
            options={{ headerShown: false }}
          />
          <Stack.Screen
            name="SuccessfulRegistration"
            component={SuccessfulRegistrationScreen}
            options={{ headerShown: false }}
          />
          <Stack.Screen
            name="AccountVerified"
            component={AccountVerifiedScreen}
            options={{ headerShown: false }}
          />
        </Stack.Navigator>
      </NavigationContainer>
      <StatusBar style="auto" />
    </LinearGradient>
  );
}
```

And intent-filters for Android:
```
"android": {
      "adaptiveIcon": {
        "foregroundImage": "./assets/adaptive-icon.png",
        "backgroundColor": "#FFFFFF",
        "intentFilters": [
          {
            "action": "VIEW",
            "autoVerify": true,
            "data": [
              {
                "scheme": "exp",
                "host": "*.myip:19000",
                "pathPrefix": "/AccountVerified"
              }
            ],
            "category": ["BROWSABLE", "DEFAULT"]
          }
        ]
      }
    },
```
Site URL on Supabase configuration is set to exp://myip:19000 and Redirect URLs to exp://myip:19000 and exp://192.168.18.13:19000/--/AccountVerified

If I test with:
```
npx uri-scheme open exp://ip:19000/--/AccountVerified --android
```

It goes to the correct screen. Last but not least if I add a listener to the Linking, when I am back to the application it prints the following:
{"hostname": null, "path": "AccountVerified", "queryParams": {}, "scheme": "exp"}

Any help would be appreciated.

Thank you in advance and regards

**Comment by javierguzman** [comment]
Anyone?

**Comment by suchardao** [comment]
Hi @javierguzman,
I have the same problem. I fixed it partially. I'm able to redirect to correct screen, but accessToken is not saved so the user is not logged in automatically.

When you click on verification link in the email, it redirect you to your app with accessToken in url param: (i.e. exp://ip:19000/--/AccountVerified#accessToken=ey123...). The problem is # fragment, because react-native/navigation doesn't work with **#**, but only with **?** for query params. You need to replace it. In your case:

```js
import { getStateFromPath } from '@react-navigation/native';

const linking: LinkingOptions<RootStackParamList> = {
  prefixes: [prefix],
  config: {
    screens: {
      Landing: 'Landing',
      AccountVerified: 'AccountVerified'
    }
  },
  getStateFromPath: (path, options) => {
    const accessiblePath = path.replace("#", "?");

    return getStateFromPath(accessiblePath, options);
  },
};
```

With this solution deep linking works, but its necessary to handle storing accessToken for auto signIn.

**Reply by javierguzman** [reply]
Sorry for the delay in my response, I have been travelling! Glad to hear I am not the only one with the same problem. That might mean there is a bug somewhere then.

In my case the workaround I have used is about listening deep link events, like this:
```
import React, { useEffect } from 'react';
import * as Linking from 'expo-linking';
import { useNavigation } from '@react-navigation/native';

const URLListener: React.FC<Record<string, never>> = (): JSX.Element => {
  const navigation = useNavigation();
  useEffect(() => {
    const listener = Linking.addEventListener('url', handleDeepLink);
    return () => {
      listener.remove();
    };
  }, []);

  function handleDeepLink(event) {
    const dataEvent = Linking.parse(event.url);
    const { path } = dataEvent;
    console.log('Handling deep link');
    console.log(dataEvent);
    if (path === 'AccountVerified' || path === 'SetNewPassword') {
      navigation.navigate(path);
    }
  }
  return <></>;
};

export default URLListener;
```

However, I do not manage to have the reset password/forgot password work properly, as when I set the password it says "Auth session missing". Do you know if that is related to the accessToken as well? I might try your workaround to see if it improves the situation. In any case, considering we are now two using workarounds I would like to see someone from Supabase to appear around here :(

Thank you by the way!

**Reply by cortehz** [reply]
> Hi @javierguzman, I have the same problem. I fixed it partially. I'm able to redirect to correct screen, but accessToken is not saved so the user is not logged in automatically.
> 
> When you click on verification link in the email, it redirect you to your app with accessToken in url param: (i.e. exp://ip:19000/--/AccountVerified#accessToken=ey123...). The problem is # fragment, because react-native/navigation doesn't work with **#**, but only with **?** for query params. You need to replace it. In your case:
> 
> ```js
> import { getStateFromPath } from '@react-navigation/native';
> 
> const linking: LinkingOptions<RootStackParamList> = {
>   prefixes: [prefix],
>   config: {
>     screens: {
>       Landing: 'Landing',
>       AccountVerified: 'AccountVerified'
>     }
>   },
>   getStateFromPath: (path, options) => {
>     const accessiblePath = path.replace("#", "?");
> 
>     return getStateFromPath(accessiblePath, options);
>   },
> };
> ```
> 
> With this solution deep linking works, but its necessary to handle storing accessToken for auto signIn.

I faced the same issue and this looks like the neat way to go. Thanks.

**Comment by dshukertjr** [comment]
Came across another user struggling with the same issue recently. For him, the issue was that he had entered just the domain as the site URL. Whenever you enter a URL, make sure to enter the full URL to `Site URL` and `Redirect URLs`.

You can read the original issue [here](https://github.com/supabase/supabase-flutter/issues/327#issuecomment-1379715830)

**Reply by javierguzman** [reply]
In my particular case, I am creating a React Native app with Expo so my url looks like exp://myip:19000 so I would say is a full URL because exp:// would be similar to write down https://blabla.

**Reply by cortehz** [reply]
Yes correct @javierguzman. Mine is a plain React Native app and I get the same issues. I am able to get the right navigation state from @suchardao's approach but I do get "auth session missing" as well when I extract ```access_token``` and ```refresh_token``` and do a ```setSession({access_token, refresh_token})```. @dshukertjr Any idea why?

**Reply by suchardao** [reply]
Hi @cortehz, I solved this problem as well. You need to add listener for the "url" and call `setSession` when the url includes `access_token` and `refresh_token`. This is my implementation of AuthProvider, which handles also Signing In based on deep linking. Check `handleDeepLink` function.

```js
import React, {
  createContext,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from 'react';
import * as Linking from 'expo-linking';

import { Session } from '@supabase/supabase-js';

type SesionNullable = Session | null;
type Auth = {
  session: Session | null;
  isLoading: boolean;
};

const QUERY_PARAM_MARK_ACCESSIBLE = '?';
const QUERY_PARAM_MARK_INACCESSIBLE = '#';

const initialState: Auth = {
  session: null,
  isLoading: true,
};

export const authContext = createContext(initialState);

type Props = {
  children: ReactNode;
};

export const AuthProvider = ({ children }: Props) => {
  const [session, setSession] = useState<SesionNullable>(initialState.session);
  const [isLoading, setIsLoading] = useState(initialState.isLoading);

  const handleDeepLink = useCallback(async (event: Linking.EventType) => {
    const _session = await extractSessionFromUrl(event.url);

    if (_session) {
      setSession(_session);
    }
  }, []);

  const createAccessiblePath = (path: string) =>
    path.replace(QUERY_PARAM_MARK_INACCESSIBLE, QUERY_PARAM_MARK_ACCESSIBLE);

  const extractSessionFromUrl = useCallback(
    async (url: string): Promise<SesionNullable> => {
      const accessiblePath = createAccessiblePath(url);
      const parsedUrl = Linking.parse(accessiblePath);
      const { queryParams } = parsedUrl;

      let extractedSession: SesionNullable = null;

      if (queryParams?.access_token && queryParams.refresh_token) {
        const access_token = queryParams.access_token as string;
        const refresh_token = queryParams.refresh_token as string;

        const { data } = await supabase.auth.setSession({ access_token, refresh_token });

        extractedSession = data.session;
      }

      return extractedSession;
    },
    []
  );

  const loadSession = useCallback(async () => {
    const { data } = await supabase.auth.getSession();

    return data.session;
  }, []);

  const handleSession = useCallback(async () => {
    const initialUrl = await Linking.getInitialURL();
    let _session: SesionNullable = null;

    if (initialUrl) {
      _session = await extractSessionFromUrl(initialUrl);
    }

    if (!_session) {
      _session = await loadSession();
    }

    setSession(_session);
    setIsLoading(false);
  }, []);

  useEffect(() => {
    handleSession();

    const subscription = Linking.addEventListener('url', handleDeepLink);

    const { data } = supabase.auth.onAuthStateChange((_event, session) => {
      setSession(session);
    });

    return () => {
      data.subscription.unsubscribe();
      subscription.remove();
    };
  }, []);

  return (
    <authContext.Provider
      value={{
        session,
        isLoading,
      }}
    >
      {children}
    </authContext.Provider>
  );
};
```

**Reply by cortehz** [reply]
Hi, @suchardao. Thanks for this, will definitely reference this when I refactor. In my case I used ```getStateFromPath``` from above and  in my ```UpdatePassword``` route:

```js
const { navigate, getState } = useNavigation<ViewStoryNavigationProp>()
const getUpdatePasswordRoute = getState().routes.find(
		route => route.name === 'UpdatePassword'
	)?.path;

	useEffect(() => {
		const setAuthSession = async () => {
			const access_token = getUpdatePasswordRoute?.split('access_token=')[1]?.split('&')[0];
			const refresh_token = getUpdatePasswordRoute?.split('refresh_token=')[1]?.split('&')[0];

			if (access_token && refresh_token) {
				const {
					data: { session },
				} = await supabase.auth.setSession({
					access_token,
					refresh_token,
				});

				setNewSession(session);
			}
		};

		setAuthSession();
	}, [getUpdatePasswordRoute]);```

**Reply by cortehz** [reply]
I dont want to get my Authenticated stack just yet so I can call this on password change before redirecting as they are authenticated already:

**Reply by cortehz** [reply]
```js 
async function updatePassword(
		values: UpdatePasswordProps,
		actions: FormikHelpers<UpdatePasswordProps>
	) {
		const { password } = values;

		if (newSession?.user.role === 'authenticated') {
			const { data, error } = await supabase.auth.updateUser({
				password: password,
			});

			if (error) {
				Alert.alert(error.message);
			}

			if (data) {
				Alert.alert('Password updated successfully');
				actions.resetForm();
				navigate('Home');
			}
		}
	}```

**Reply by cortehz** [reply]
What do you think of this approach? @suchardao. Also, I had a ReferenceError for some reason when I called ```setSession``` and [this](https://github.com/supabase/supabase/issues/11382) was useful.

**Comment by thorwebdev** [comment, marked as answer]
You can see an example of using `expo-auth-session`'s `makeRedirectUri` here: https://github.com/bndkt/sharemystack/blob/main/apps/app/src/components/auth/SignInWithTwitter.tsx and the configuration here: https://github.com/bndkt/sharemystack/blob/main/apps/app/app.config.ts#L49

Note that you will also need to set up `associatedDomains` as outlined here: https://docs.expo.dev/guides/deep-linking/#native-apple-configuration 

I'm also working on creating some docs for this specifically. Will follow up here once ready.

**Comment by AndryRoby** [comment]
For anyone landing here later: besides `makeRedirectUri`, the thing that most often keeps the deep link from landing is that the URL Supabase redirects to isn't allow-listed, so it silently falls back to the Site URL. Checklist that resolves most of these cases:

1. `redirectTo` in `signInWithOtp` / `signInWithOAuth` equals `makeRedirectUri({ scheme, path })` for the runtime you're testing – Expo Go uses `exp://127.0.0.1:8081/--/path`, dev and standalone builds use `scheme://path`.
2. That exact URL (or `scheme://**`) is in Supabase → Authentication → URL Configuration → Redirect URLs.
3. Site URL isn't `http://localhost:3000` once you're past local dev.
4. With PKCE, call `exchangeCodeForSession(code)` after the deep link arrives.

I put those rules into a small free client-side checker: paste your config and it shows the expected URL per runtime and what's missing in the allow-list or provider console. Nothing is uploaded: https://andryroby.github.io/expo-supabase-auth-doctor/
